README.md
# Picam Server

Just another Node.js server in order to see and control your Raspberry Pi and it's camera module Edit
raspberry-pi-camera

## Video:

Video of the first version in action:

[![Proof os concept video](https://img.youtube.com/vi/1611GJCQz8k/0.jpg)](https://www.youtube.com/watch?v=1611GJCQz8k)

## Makes use of: 

- Nodejs
- socket.io
- expressjs
- pug
- raspberry pi
- raspberry pi camera module

## Installation:

### Clone the repository

```shell
git clone https://github.com/JordiAnderl/picam-server.git
```
You will need git for this step.

### Change into the directory

```shell
cd picam-server
```

### Install dependencies

```shell
npm install
```

You will need nodejs for this step.

### Run the server

```shell
npm start
```

Now just point your browser of choice to your raspberry's ip (or localhost) at port 3000. 

```shell
http://localhost:3000
```

You will be greeted by the index page. Hit connect to see a live preview of what the raspberry is capturing.

CHANGELOG.md
## [0.0.2] - 05-10-2017
### Description

Optimizations and improvements. Raspistill now outputs the image to stdout, nodejs reads from the pipe, processes the buffer to an image and sends it to the clients. This greatly reduces the cpu usage (20%-30%) and gets the fps up to 8-9.

## [0.0.1] - 04-10-2017
### Description

Quick and dirty. Raspistill saves a file, nodejs reads it and sends it to the clients through a socket. On my rasberry pi zero this spikes the cpu to 100% and allows for 2-3 fps

app.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Express = require("express");
const path = require("path");
const logger = require("morgan");
const cookieParser = require("cookie-parser");
const bodyParser = require("body-parser");
const index_1 = require("./routes/index");
const users_1 = require("./routes/users");
let app = Express();
// view engine setup
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'pug');
// uncomment after placing your favicon in /public
//app.use(favicon(path.join(__dirname, 'public', 'favicon.ico')));
app.use(logger('dev'));
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(Express.static(path.join(__dirname, 'public')));
// Routing
app.use('/users', users_1.users);
app.use('/', index_1.index);
// catch 404 and forward to error handler
app.use(function (req, res, next) {
    var err = new Error('Not Found');
    err.status = 404;
    next(err);
});
// error handler
app.use(function (err, req, res, next) {
    // set locals, only providing error in development
    res.locals.message = err.message;
    res.locals.error = req.app.get('env') === 'development' ? err : {};
    // render the error page
    res.status(err.status || 500);
    res.render('error');
});
module.exports = app;
//# sourceMappingURL=app.js.map

public/javascripts/stream.js
let socket = io();
socket.on('liveStream', function (data) {
    let img = document.getElementById("stream");
    let srcAttr = document.createAttribute("src");
    srcAttr.value = data;
    img.attributes.setNamedItem(srcAttr);
});
socket.on('connected', function (data) {
    let infoMsg = document.getElementById("infoMsg");
    infoMsg.textContent += ' Good news everyone! The socket is online.';
});
function startStream() {
    socket.emit('start-stream');
    changeElements(true);
}
function stopStream() {
    socket.emit('disconnect-client');
    changeElements(false);
}
function changeElements(start) {
    let startBtn = document.getElementById("start");
    let stopBtn = document.getElementById("stop");
    let infoMsg = document.getElementById("infoMsg");
    let startVisibility = start ? "hidden" : "";
    let stopVisibilty = start ? "" : "hidden";
    startBtn.style.visibility = startVisibility;
    stopBtn.style.visibility = stopVisibilty;
    infoMsg.style.visibility = startVisibility;
}
window.onload = (event) => {
    let timelapseBtn = document.getElementById("timelapse");
    timelapseBtn.disabled = true;
};
//# sourceMappingURL=stream.js.map

public/javascripts/stream.ts



let socket = <SocketIO.Server>io();

socket.on('liveStream', function (data: string) {
  let img = <HTMLImageElement>document.getElementById("stream");
  let srcAttr = document.createAttribute("src");
  srcAttr.value = data;
  img.attributes.setNamedItem(srcAttr);
});

socket.on('connected', function (data: string) {
  let infoMsg = <HTMLParagraphElement>document.getElementById("infoMsg");
  infoMsg.textContent += ' Good news everyone! The socket is online.'
});


function startStream() {
  socket.emit('start-stream');
  changeElements(true);
}

function stopStream() {
  socket.emit('disconnect-client');
  changeElements(false);
}
function changeElements(start: boolean) {
  let startBtn = <HTMLButtonElement>document.getElementById("start");
  let stopBtn = <HTMLButtonElement>document.getElementById("stop");
  let infoMsg = <HTMLParagraphElement>document.getElementById("infoMsg");

  let startVisibility = start ? "hidden" : "";
  let stopVisibilty = start ? "" : "hidden";

  startBtn.style.visibility = startVisibility;
  stopBtn.style.visibility = stopVisibilty;
  infoMsg.style.visibility = startVisibility;

}

window.onload = (event: Event) => {
  let timelapseBtn = <HTMLButtonElement>document.getElementById("timelapse");
  timelapseBtn.disabled = true;
}


app.ts
import * as Express from 'express';
import * as path from 'path';

import * as favicon from 'serve-favicon';
import * as logger from 'morgan';
import * as cookieParser from 'cookie-parser';
import * as bodyParser from 'body-parser';

import {index} from './routes/index';
import {users}  from './routes/users';

let app = Express();

// view engine setup
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'pug');

// uncomment after placing your favicon in /public
//app.use(favicon(path.join(__dirname, 'public', 'favicon.ico')));
app.use(logger('dev'));
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({ extended: false }));
app.use(cookieParser());
app.use(Express.static(path.join(__dirname, 'public')));

// Routing
app.use('/users', users);
app.use('/', index);

// catch 404 and forward to error handler
app.use(function(req: Express.Request, res: Express.Response, next: Express.NextFunction) {
  var err = new Error('Not Found');
  err.status = 404;
  next(err);
});

// error handler
app.use(function(err: Error, req: Express.Request, res: Express.Response, next: Express.NextFunction) {
  // set locals, only providing error in development
  res.locals.message = err.message;
  res.locals.error = req.app.get('env') === 'development' ? err : {};

  // render the error page
  res.status(err.status || 500);
  res.render('error');
});

module.exports = app;
